### PR TITLE
Fix multiple blocking binary reader issues

### PR DESCRIPTION
**Pull request by nirosys**
*Issue #, if available:*

*Description of changes:*

- Fixes an issue where `read_from` would read fewer bytes than it was expected to.
- Addded incomplete guard to an annotation consume that could cause an error when the buffer did not have enough data.
- Fixed an issue where bytes read into a NBR were counted twice when determining if all data had been read. This probably wasn't an issue since `read_from` reads all of the data requested.
- Implements stream completion for non-blocking binary reader.
- Moves Expandable to the raw_reader module now that the binary raw reader is using it.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by zslayton — COMMENTED** [review]
The diff looks good! Could you add unit tests that cover the corner cases this addresses?

**Comment by nirosys** [comment]
> The diff looks good! Could you add unit tests that cover the corner cases this addresses?

Absolutely, I'll try to push those up tomorrow.

**Comment by nirosys** [comment]
Pushed up some unit tests:

### `read_from` reading fewer bytes, and potentially panicking.
Pre-PR:
```
thread 'binary::non_blocking::binary_buffer::tests::validate_read_from_size' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `11`', src/binary/non_blocking/binary_buffer.rs:821:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test binary::non_blocking::binary_buffer::tests::validate_read_from_size ... FAILED

failures:

failures:
    binary::non_blocking::binary_buffer::tests::validate_read_from_size
```
Post PR:
```
test binary::non_blocking::binary_buffer::tests::validate_read_from_size ... ok
```

### Processing annotation bytes leading to panic from BinaryBuffer over-consume.
Pre-PR:
```
thread 'binary::non_blocking::raw_binary_reader::tests::test_incomplete_annotation_wrapper' panicked at 'assertion failed: num_bytes_to_consume <= self.remaining()', src/binary/non_blocking/binary_buffer.rs:121:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test binary::non_blocking::raw_binary_reader::tests::test_incomplete_annotation_wrapper ... FAILED

failures:

failures:
    binary::non_blocking::raw_binary_reader::tests::test_incomplete_annotation_wrapper
```
Post PR:
```
test binary::non_blocking::raw_binary_reader::tests::test_incomplete_annotation_wrapper ... ok
```

### End of Stream implementation in binary reader.
Pre-PR:
```
thread 'blocking_reader::tests::test_binary_end_of_stream' panicked at 'assertion failed: `(left == right)`
  left: `Value(Bool)`,
 right: `Nothing`', src/blocking_reader.rs:561:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test blocking_reader::tests::test_binary_end_of_stream ... FAILED

failures:

failures:
    blocking_reader::tests::test_binary_end_of_stream
```

Post-PR:
```
test blocking_reader::tests::test_binary_end_of_stream ... ok
```

File: src/binary/non_blocking/binary_buffer.rs [hunk 1]
```diff
@@ -808,4 +808,17 @@ mod tests {
             .expect_err("This exceeded the configured max Int size.");
         Ok(())
     }
+
+    #[test]
+    fn validate_read_from_size() -> IonResult<()> {
+        // This test validates that the size of data we wish to read, is actually honored by
```
Comment (zslayton) [inline src/binary/non_blocking/binary_buffer.rs:814]:
```suggestion
        // This test validates that the size of data we wish to read is actually honored by
```

**Review by zslayton — APPROVED** [review]
Glad to have this fixed, thanks!
